File: tools/aseprite/verify_toolchain.py
"""Run real Aseprite Lua and CLI export checks, without touching game assets."""
from __future__ import annotations

import argparse
import hashlib
import json
from pathlib import Path
import subprocess

from PIL import Image


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--executable", type=Path, default=Path("F:/Tools/Aseprite/chuwuzhi-v1.3.18.5/build/bin/aseprite.exe"))
    parser.add_argument("--output", type=Path, default=Path("artifacts/aseprite-toolchain"))
    args = parser.parse_args()
    output = args.output.resolve()
    output.mkdir(parents=True, exist_ok=True)
    script = Path(__file__).with_suffix(".lua").resolve()
    creation = subprocess.run([str(args.executable), "--batch", "--script-param", f"output={output.as_posix()}", "--script", str(script)], check=True, capture_output=True, text=True)
    assert "ASEPRITE_TOOLCHAIN_FIXTURE_OK" in creation.stdout, creation.stdout
    subprocess.run([str(args.executable), "--batch", str(output / "fixture.aseprite"), "--sheet-type", "horizontal", "--sheet", str(output / "fixture.png"), "--data", str(output / "fixture.json"), "--format", "json-array", "--list-layers", "--list-tags", "--list-slices"], check=True)
    data = json.loads((output / "fixture.json").read_text(encoding="utf-8"))
    assert len(data["frames"]) == 2
    assert [frame["duration"] for frame in data["frames"]] == [100,150]
    assert [layer["name"] for layer in data["meta"]["layers"]] == ["body","detail"]
    assert data["meta"]["frameTags"][0]["name"] == "fixture_loop"
    assert data["meta"]["slices"][0]["keys"][0]["pivot"] == {"x":8,"y":14}
    image = Image.open(output / "fixture.png").convert("RGBA")
    assert image.size == (32,16), image.size
    assert {alpha for *_,alpha in image.get_flattened_data()} == {0,255}
    colors = {pixel[:3] for pixel in image.get_flattened_data() if pixel[3]}
    assert colors == {(44,39,47),(209,154,75),(247,230,194)}, colors
    # The only difference between frames is one intended highlight pixel.
    before = image.crop((0,0,16,16))
    after = image.crop((16,0,32,16))
    assert sum(a != b for a,b in zip(before.get_flattened_data(),after.get_flattened_data())) == 1
    hashes = {path.name:hashlib.sha256(path.read_bytes()).hexdigest() for path in output.iterdir() if path.suffix in {".aseprite",".png",".json"} and path.name != "report.json"}
    report = {"status":"passed", "method":"real Aseprite batch Lua, save/reopen, CLI spritesheet export, decoded PNG and JSON inspection", "fixture_only":True, "art_quality_validated":False, "lua_output":creation.stdout.strip(), "checks":["indexed palette preserved", "two editable layers preserved", "two frames and exact durations", "animation tag preserved", "slice and feet pivot preserved", "transparent PNG alpha is binary", "exact three visible colors", "exactly one intended pixel changes across frames"], "files":hashes}
    (output / "report.json").write_text(json.dumps(report,indent=2),encoding="utf-8")
    print(json.dumps(report))


if __name__ == "__main__":
    main()

File: tools/aseprite/verify_toolchain.lua
-- Synthetic toolchain fixture. This is not game artwork or an art-quality test.
local output = assert(app.params["output"], "Pass --script-param output=<directory>")
local sprite = Sprite(16, 16, ColorMode.INDEXED)
local palette = Palette(4)
palette:setColor(0, Color{r=0, g=0, b=0, a=0})
palette:setColor(1, Color{r=44, g=39, b=47, a=255})
palette:setColor(2, Color{r=209, g=154, b=75, a=255})
palette:setColor(3, Color{r=247, g=230, b=194, a=255})
sprite:setPalette(palette)
sprite.transparentColor = 0

local body = sprite.layers[1]
body.name = "body"
local detail = sprite:newLayer()
detail.name = "detail"
local base = Image(16, 16, ColorMode.INDEXED)
for y=5,13 do
  for x=4,11 do base:drawPixel(x,y,1) end
end
sprite:newCel(body, 1, base, Point(0,0))
local mark = Image(16, 16, ColorMode.INDEXED)
for y=6,9 do
  for x=6,9 do mark:drawPixel(x,y,2) end
end
sprite:newCel(detail, 1, mark, Point(0,0))
sprite:newFrame(1)
local moved = mark:clone()
moved:drawPixel(7,6,3)
sprite:newCel(detail, 2, moved, Point(0,0))
sprite.frames[1].duration = 0.1
sprite.frames[2].duration = 0.15
local tag = sprite:newTag(1,2)
tag.name = "fixture_loop"
local anchor = sprite:newSlice(Rectangle(0,0,16,16))
anchor.name = "feet"
anchor.pivot = Point(8,14)
anchor.data = "toolchain-test"
sprite:saveAs(output .. "/fixture.aseprite")
sprite:close()
local reopened = assert(app.open(output .. "/fixture.aseprite"))
assert(#reopened.layers == 2, "layers not preserved")
assert(#reopened.frames == 2, "frames not preserved")
assert(reopened.tags[1].name == "fixture_loop", "tag not preserved")
assert(reopened.slices[1].pivot.x == 8 and reopened.slices[1].pivot.y == 14, "pivot not preserved")
assert(reopened.palettes[1]:getColor(2).red == 209, "palette not preserved")
print("ASEPRITE_TOOLCHAIN_FIXTURE_OK " .. tostring(app.version) .. " api=" .. app.apiVersion)
